File: py3动态规划/线性DP/AcWing 899. 编辑距离.py
from bisect import bisect_left, bisect_right, insort_left, insort_right, insort, bisect
from math import ceil, floor, pow, gcd, sqrt, log10, fabs, fmod, factorial, inf, pi, e
from heapq import heapify, heapreplace, heappush, heappop, heappushpop, nlargest, nsmallest
from typing import List, Tuple, Optional, Dict, Set
from collections import defaultdict, Counter, deque, OrderedDict, namedtuple
from itertools import permutations, combinations, combinations_with_replacement, accumulate, count, groupby
from queue import PriorityQueue, Queue, LifoQueue
from functools import lru_cache, reduce

MOD = int(1e9 + 7)
INF = int(1e20)
INFMIN = float('-inf')
INFMAX = float('inf')

# -*- coding: utf-8 -*-
# @Author  : wheat
# @Time    : 2023/02/01 17:30

if __name__ == '__main__':
    N = 15
    M = 1010
    f = [[0] * N for _ in range(N)]
    a = [[0] * N for _ in range(M)]

    def distance(a, b):
        la = len(a) - 1
        lb = len(b) - 1

        for i in range(1, lb + 1):
            f[0][i] = i
        for i in range(1, la + 1):
            f[i][0] = i

        for i in range(1, la + 1):
            for j in range(1, lb + 1):
                f[i][j] = min(f[i - 1][j] + 1, f[i][j - 1] + 1)
                if a[i] == b[j]:
                    f[i][j] = min(f[i][j], f[i - 1][j - 1])
                else:
                    f[i][j] = min(f[i][j], f[i - 1][j - 1] + 1)
        return f[la][lb]

    n, m = map(int, input().split())
    for i in range(n):
        a[i] = " " + input()

    while m:
        m -= 1
        li = input().split()
        b = " " + li[0]
        lmt = int(li[1])
        res = 0
        for i in range(n):
            if distance(a[i], b) <= lmt:
                res += 1
        print(res)

File: C++语法/C++动态规划/线性DP/AcWing 899. 编辑距离.cpp
#include <bits/stdc++.h>

using namespace std;

/*
    -*- coding: utf-8 -*-
    @Author  : wheat
    @Time    : 2023/02/01 17:32
*/


const int N = 11, M = 1001;
int n, m, f[N][N];
char str[M][N];

int distans(char a[], char b[]) {
    int l1 = strlen(a + 1), l2 = strlen(b + 1);
    for (int i = 1; i <= l1; i++)
        for (int j = 1; j <= l2; j++) {
            f[i][j] = min(f[i - 1][j], f[i][j - 1]) + 1;
            if (a[i] == b[j]) f[i][j] = min(f[i][j], f[i - 1][j - 1]);
            else f[i][j] = min(f[i][j], f[i - 1][j - 1] + 1);
        }
    return f[l1][l2];
}

int main() {
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= 10; i++) f[i][0] = f[0][i] = i;
    for (int i = 0; i < n; i++) scanf("%s", str[i] + 1);
    while (m--) {
        char s[N];
        int limit, res = 0;
        scanf("%s%d", s + 1, &limit);
        for (int i = 0; i < n; i++)
            res += distans(str[i], s) <= limit;
        printf("%d\n", res);
    }
}
